File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_02_arithmetic/P39.java
package com.shekhargulati.ninetynine_problems._02_arithmetic;

import java.util.stream.IntStream;

/**
 * (**) Calculate Euler's totient function phi(m).
 */
public class P39 {

    public static long phi(int m) {
        return IntStream.rangeClosed(1, m).filter(r -> P38.coprime(r, m)).count();
    }
}

File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_02_arithmetic/P39.scala
package com.shekhargulati.ninetynine_problems._02_arithmetic

/**
  * 2.09 (**) Calculate Euler's totient function phi(m).
  */
object P39 {

  def totient_phi(m: Int): Int = (1 to m).count(r => P38.coprime(m, r))


  /**
    * The cototient of n is defined as n – φ(n), i.e. the number of positive integers less than or equal to n that are divisible by at least one prime that also divides n.
    *
    * @param m number
    * @return cototient
    */
  def cototient(m: Int): Int = m - totient_phi(m)

}
